tests/storages_tests/test_storages.py

```diff
@@ -711,13 +881,112 @@ def _create_new_trial_with_example_trial(storage, study_id, distributions, examp
+                        best_trial = trial
+        study_id_to_trials[study_id] = trials
+        study_id_to_summary[study_id] = StudySummary(
+            study_name=study_name,
+            direction=direction,
+            best_trial=best_trial,
+            user_attrs={},
+            system_attrs={},
+            n_trials=len(trials),
+            datetime_start=datetime_start,
+            study_id=study_id,
+        )
+    return study_id_to_summary, study_id_to_trials
+
+
+def _generate_trial(generator: random.Random) -> FrozenTrial:
+    example_params = {
+        "paramA": (generator.uniform(0, 1), UniformDistribution(0, 1)),
+        "paramB": (generator.uniform(1, 2), LogUniformDistribution(1, 2)),
+        "paramC": (
+            generator.choice(["CatA", "CatB", "CatC"]),
+            CategoricalDistribution(["CatA", "CatB", "CatC"]),
+        ),
+        "paramD": (generator.uniform(-3, 0), UniformDistribution(-3, 0)),
+        "paramE": (generator.choice([0.1, 0.2]), CategoricalDistribution([0.1, 0.2])),
+    }
+    example_attrs = {
+        "attrA": "valueA",
+        "attrB": 1,
+        "attrC": None,
+        "attrD": {"baseline_score": 0.001, "tags": ["image", "classification"]},
+    }
+    state = generator.choice(ALL_STATES)
+    params = {}
+    distributions = {}
+    user_attrs = {}
+    system_attrs = {}
+    intermediate_values = {}
+    for key, (value, dist) in example_params.items():
+        if generator.choice([True, False]):
+            params[key] = value
+            distributions[key] = dist
+    for key, value in example_attrs.items():
+        if generator.choice([True, False]):
+            user_attrs["usr_" + key] = value
+        if generator.choice([True, False]):
+            system_attrs["sys_" + key] = value
+    for i in range(generator.randint(4, 10)):
+        if generator.choice([True, False]):
+            intermediate_values[i] = generator.uniform(-10, 10)
+    return FrozenTrial(
+        number=0,  # dummy
+        state=state,
+        value=generator.uniform(-10, 10),
+        datetime_start=datetime.now(),
+        datetime_complete=datetime.now() if state.is_finished() else None,
+        params=params,
+        distributions=distributions,
+        user_attrs=user_attrs,
+        system_attrs=system_attrs,
+        intermediate_values=intermediate_values,
+        trial_id=0,  # dummy
     )
+
+
+def _check_trial_equality(output: FrozenTrial, expected: FrozenTrial) -> None:
+    assert output._trial_id == expected._trial_id
+    assert output.state == expected.state
+    assert output.value == expected.value
+    assert output.datetime_start == expected.datetime_start
+    assert output.datetime_complete == expected.datetime_complete
+    assert output.user_attrs == expected.user_attrs
+    assert output.system_attrs == expected.system_attrs
+    assert output.intermediate_values == expected.intermediate_values
```

Comment:
`assert output == expected` is insufficient even if `FrozenTrial` has the `__eq__` method?

Reply:
It seems I have misunderstood somethin, and I replaced them with `__eq__`.

Reply:
I found that for `CategoricalDistribution`, `json_to_distribution(distribution_to_json(dist)) != dist` in general and I'll create a PR to add an appropriate __eq__ method to `CategoricalDistribution`.

Reply:
> I'll create a PR to add an appropriate eq method to CategoricalDistribution.

Good catch! Thank you.